### ISSUE TITLE
Folder as Workspace Delete key should remove "top-level" folder

### ISSUE THREAD

**Issue by vinsworldcom** [notepad-plus-plus/notepad-plus-plus#8960]
### Description of the Issue
Right-click on a top-level folder in Folder as Workspace presents a "Remove" option in the pop-up menu.  Pressing the Delete key should have the same behavior on a top-level folder.

### Expected Behavior
Press the delete key with a top-level folder selected in the Folder as Workspace dockable should remove the folder - similar to a right-click and selecting "Remove" from the pop-up menu.

### Actual Behavior
Press the delete key with a top-level folder selected in the Folder as Workspace dockable does nothing.

### Debug Information
Notepad++ v7.9   (64-bit)
Build time : Oct  4 2020 - 22:36:46
Path : C:\usr\bin\npp64\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1909
OS Build : 18363.1082
Current ANSI codepage : 1252
Plugins:

**Comment by sasumner** [comment]
Should it have a "confirm"?
It is a lot easier to accidentally hit the Delete key than it is to accidentally trigger the menu item.
And/or, perhaps input focus isn't where the user thinks when they press Delete.

**Comment by ArkadiuszMichalski** [comment]
Note that for Project1/2/3 there is confirmation.

![a](https://user-images.githubusercontent.com/2730894/95197293-aa379880-07d9-11eb-8a56-70deba55ff15.gif)

It should be here too.

**Comment by vinsworldcom** [comment]
**_I_** don't think it should have a confirm, but I'm not against that.

**NOTE**: just saw reply from @ArkadiuszMichalski pop-up above while I'm typing my response.  If that is how Project Panel works, it would be fine to keep consistent.

I see that `VK_DELETE` is [already in the fileBrowser.cpp code](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8426c9ccd98157d2712e1b54d54f498a32e6481f/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp#L654), just commented out.  It triggers `IDM_FILEBROWSER_DELETEFOLDER` - not sure if that is for _any_ folder or just the top-level ones like I'm advocating.

I don't think the Folder as Workspace is meant as a "file manager" to actually add and delete files / folders on disk; rather, just add / remove the top-level folders in the Folder as Workspace dockable.

Cheers.

**Comment by ArkadiuszMichalski** [comment]
Yes, both are not "file manager" and don't remove files from disk, but other case works for this panels (like Home/End/PageUP/PageDown) so pressing Delete by mistake may occure.

**Comment by vinsworldcom** [comment]
@ArkadiuszMichalski good point.  And again, I'm not opposed to adding a confirm message box to keep consistent with Project Panel.  Also note I'm only advocating that Delete work on "top-level" folders in Folder as Workspace, not sub folders and certainly not files.  Pressing Delete on anything but a top-level folder should do nothing (or issue that lovely system beep :-)

To be clear (as one can be with ASCII art), Delete should only prompt for deletion when pressed on "MyTopLevelFolder1" and "MyTopLevelFolder2" shown below:

```
+ MyTopLevelFolder1
|--subFolder1
|--subFolder2
|----FileInSub21
|----FileInSub22
|--FileInTopLevel11
+ MyTopLevelFolder1
|--FileInTopLevel21
```

Any other Delete key presses should do nothing.  In this way, we're just replicating on the keyboard, the option "Remove" we have from the pop-up menu.

Cheers.

**Comment by ArkadiuszMichalski** [comment]
Sure, it's obviouse that you mean only root, even from context menu you can't delete subfolder. Root is sync with real root's content on disk (when you add/change files/subfolders on disk then this change will apper in list).

**Comment by vinsworldcom** [comment]
Just tested "hijacking" the commented out `VK_DELETE` with slight modification, and it works to delete only top-level folders.  

**NOTE:**  This does **_not_** do any confirmation dialogue. I looked at the [Project Panel confirm dialogue](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/8426c9ccd98157d2712e1b54d54f498a32e6481f/PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp#L1111) and it looks to just do a MessageBox, but connects to the "native language" translations and I'm not sure how to include that for anything other than English.  Also, the Project Panel confirm dialogue happens in the actual `case` statement for the `IDM_` message, so it probably triggers on both the Delete key and the pop-up menu remove.  Not sure how to add the confirm here:

- Just for the Delete key
- change the current behavior of the Remove pop-up menu choice and have that prompt as well


`git diff` follows (without the confirmation:

```
PS VinsWorldcom ~\source\personal\notepad-plus-plus\PowerEditor\visual.net > git diff ..\src\WinControls\FileBrowser\fileBrowser.cpp
diff --git a/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp b/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
index 1907c843..97d6b8e3 100644
--- a/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
+++ b/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
@@ -650,16 +650,16 @@ void FileBrowser::notified(LPNMHDR notification)
                                        else
                                                _treeView.toggleExpandCollapse(hItem);
                                }
-                               /*
                                else if (ptvkd->wVKey == VK_DELETE)
                                {
                                        HTREEITEM hItem = _treeView.getSelection();
                                        BrowserNodeType nType = getNodeType(hItem);
-                                       if (nType == browserNodeType_folder)
-                                               popupMenuCmd(IDM_FILEBROWSER_DELETEFOLDER);
-                                       else if (nType == browserNodeType_file)
-                                               popupMenuCmd(IDM_FILEBROWSER_DELETEFILE);
+                                       if (nType == browserNodeType_root)
+                                               popupMenuCmd(IDM_FILEBROWSER_REMOVEROOTFOLDER);
                                }
+                               /*
                                else if (ptvkd->wVKey == VK_UP)
                                {
                                        if (0x80 & GetKeyState(VK_CONTROL))
```

**Comment by vinsworldcom** [comment]
Now with the message box confirm just for the Delete key - not the "Remove" pop-up menu item.  And not using Localization .. it works as expected:

```
PS VinsWorldcom ~\source\personal\notepad-plus-plus\PowerEditor\visual.net > git diff ..\src\WinControls\FileBrowser\fileBrowser.cpp
diff --git a/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp b/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
index 1907c843..199b744b 100644
--- a/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
+++ b/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp
@@ -650,16 +650,22 @@ void FileBrowser::notified(LPNMHDR notification)
                                        else
                                                _treeView.toggleExpandCollapse(hItem);
                                }
-                               /*
                                else if (ptvkd->wVKey == VK_DELETE)
                                {
                                        HTREEITEM hItem = _treeView.getSelection();
                                        BrowserNodeType nType = getNodeType(hItem);
-                                       if (nType == browserNodeType_folder)
-                                               popupMenuCmd(IDM_FILEBROWSER_DELETEFOLDER);
-                                       else if (nType == browserNodeType_file)
-                                               popupMenuCmd(IDM_FILEBROWSER_DELETEFILE);
+                                       if (nType == browserNodeType_root)
+                                       {
+                                               int res = ::MessageBox(
+                                                       _hSelf,
+                                                       TEXT("All the sub-items will be removed.\rAre you sure you want to remove this folder from the workspace?"),
+                                                       TEXT("Remove folder from workspace"),
+                                                       MB_YESNO);
+                                               if (res == IDYES)
+                                                       popupMenuCmd(IDM_FILEBROWSER_REMOVEROOTFOLDER);
+                                       }
                                }
+                               /*
                                else if (ptvkd->wVKey == VK_UP)
                                {
                                        if (0x80 & GetKeyState(VK_CONTROL))
```

Thoughts on:

- Whether the confirm message box should be for **both** `VK_DELETE` **and** "Remove" menu item, or just `VK_DELETE`?
- Should this change / PR include the localization updates required to make this translatable - meaning I'd need to update at least "english.xml" and "english_customizable.xml" .. what to do about other localization XML files?
    - Should this change / PR only be the `VK_DELETE` and a separate change / PR to make it localized? 


Comments?

Cheers.

**Comment by ArkadiuszMichalski** [comment]
In my opinion both (FAW/Project) should behave the same (for Delete always confirm dialog, for context menu as you decide, but now Project also dispaly this confirm dialog). Adding translation of this box is also a good idea so that you don't come back to it later. 

@donho any suggestions?

**Comment by donho** [comment]
For me, why not DELETE key. But confirmation of removing makes DELETE key pointless.

The reason of a confirmation before removing in project panel is because the project data is user's data - removing the root of project will destroy all the tree structure that user has built in Notepad++. Whereas the tree structure in FaW is ready-only in Notepad++, it's maintained outside and Notepad++ has no responsibility to it.

So it's no big deal if user removes a root by accident - he/she can just add it back.

**Comment by vinsworldcom** [comment]
@donho 

Good point.  The PR I submitted does add the confirmation for both Delete key and the Remove menu item to act like Project Panel.  It'd be easy to cancel and submit a new PR that just add Delete key function - it's actually easier all around, since adding the confirmation I needed to add / update file localizations for translating the message.

I'll wait a bit to see if anyone wants to debate / discuss before creating the new Delete key only / no confirm PR.

Cheers.
